// fusion-core/dist-node-cjs/tokens.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EnableMiddlewareTimingToken = exports.CriticalChunkIdsToken = exports.RoutePrefixToken = exports.SSRBodyTemplateToken = exports.HttpServerToken = exports.SSRDeciderToken = exports.ElementToken = exports.RenderToken = exports.RouteTagsToken = void 0;

var _createToken = require("./create-token");

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
const RouteTagsToken = (0, _createToken.createToken)('RouteTagsToken');
exports.RouteTagsToken = RouteTagsToken;
const RenderToken = (0, _createToken.createToken)('RenderToken');
exports.RenderToken = RenderToken;
const ElementToken = (0, _createToken.createToken)('ElementToken');
exports.ElementToken = ElementToken;
const SSRDeciderToken = (0, _createToken.createToken)('SSRDeciderToken');
exports.SSRDeciderToken = SSRDeciderToken;
const HttpServerToken = (0, _createToken.createToken)('HttpServerToken');
exports.HttpServerToken = HttpServerToken;
const SSRBodyTemplateToken = (0, _createToken.createToken)('SSRBodyTemplateToken');
exports.SSRBodyTemplateToken = SSRBodyTemplateToken;
const RoutePrefixToken = (0, _createToken.createToken)('RoutePrefixToken');
exports.RoutePrefixToken = RoutePrefixToken;
const CriticalChunkIdsToken = (0, _createToken.createToken)('CriticalChunkIdsToken');
exports.CriticalChunkIdsToken = CriticalChunkIdsToken;
const EnableMiddlewareTimingToken = (0, _createToken.createToken)('EnableMiddlewareTimingToken');
exports.EnableMiddlewareTimingToken = EnableMiddlewareTimingToken;
