// fusion-react/dist-browser-esm/context.js
/** Copyright (c) 2019 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
import * as React from 'react';
import { jsx as _jsx } from "react/jsx-runtime";
export const FusionContext = /*#__PURE__*/React.createContext({});
export const ServiceContext = /*#__PURE__*/React.createContext(() => {
  throw new Error('`ServiceContext.Provider` was not found. This occurs if you are attempting to use `ServiceContext` in a non-React Fusion.js application.');
});
export function useService(token) {
  const getService = React.useContext(ServiceContext);
  const provides = getService(token);
  return provides;
}
export function ServiceConsumer({
  token,
  children
}) {
  return /*#__PURE__*/_jsx(ServiceContext.Consumer, {
    children: getService => {
      const provides = getService(token);
      return children(provides);
    }
  });
}

function getServices(getService, deps) {
  const services = {};
  Object.keys(deps).forEach(name => {
    services[name] = getService(deps[name]);
  });
  return services;
}

const identity = i => i;

export function withServices(deps, mapServicesToProps = identity) {
  function resolve(getService) {
    const services = getServices(getService, deps);
    const serviceProps = mapServicesToProps(services);
    return serviceProps;
  }

  return Component => {
    return function WithServices(props) {
      return /*#__PURE__*/_jsx(ServiceContext.Consumer, {
        children: getService => /*#__PURE__*/_jsx(Component, { ...resolve(getService),
          ...props
        })
      });
    };
  };
}
