// fusion-react/dist-browser-esm/async/prepare-provider.js
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
import * as React from 'react';
import PropTypes from 'prop-types';

class PrepareProvider extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.splitComponentLoaders = [];
    this.markAsCritical = props.markAsCritical;
    this.pushSSRMetadata = props.pushSSRMetadata;
  }

  getChildContext() {
    return {
      splitComponentLoaders: this.splitComponentLoaders,
      markAsCritical: this.markAsCritical,
      pushSSRMetadata: this.pushSSRMetadata
    };
  }

  render() {
    return React.Children.only(this.props.children);
  }

}

_defineProperty(PrepareProvider, "childContextTypes", {
  splitComponentLoaders: PropTypes.array.isRequired,
  markAsCritical: PropTypes.func,
  pushSSRMetadata: PropTypes.func
});

export default PrepareProvider;
