// fusion-core/dist-browser-cjs/plugins/timing.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.TimingToken = void 0;

var _createPlugin = require("../create-plugin");

var _memoize = require("../memoize");

var _createToken = require("../create-token");

var _now = require("../utils/now.js");

/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */
class Timing {
  constructor() {
    this.start = (0, _now.now)();
    this.render = deferred();
    this.end = deferred();
    this.downstream = deferred();
    this.upstream = deferred();
    this.upstreamStart = -1;
    this.middleware = [];
    this.prepass = [];
    this.prepassMarked = false;
    this.prepassStart = -1;
  }

  markPrepass(pendingSize) {
    if (!this.prepassMarked) {
      this.prepassMarked = true;
      this.prepassStart = (0, _now.now)();
    } else {
      this.prepass.push({
        duration: (0, _now.now)() - this.prepassStart,
        pendingSize: pendingSize || 0
      });
      this.prepassMarked = false;
      this.prepassStart = -1;
    }
  }

}

const timing = {
  from: (0, _memoize.memoize)(() => new Timing())
};
const TimingToken = (0, _createToken.createToken)('TimingToken');
exports.TimingToken = TimingToken;

function middleware(ctx, next) {
  ctx.memoized = new Map();
  const {
    start,
    render,
    end,
    downstream,
    upstream,
    middleware,
    prepass,
    markPrepass
  } = timing.from(ctx);
  ctx.timing = {
    start,
    render: render.promise,
    end: end.promise,
    downstream: downstream.promise,
    upstream: upstream.promise,
    middleware,
    prepass,
    markPrepass
  };
  return next().then(() => {
    const upstreamTime = (0, _now.now)() - timing.from(ctx).upstreamStart;
    upstream.resolve(upstreamTime);
    const endTime = (0, _now.now)() - ctx.timing.start;
    end.resolve(endTime);
  }).catch(e => {
    // currently we only resolve upstream and downstream when the request does not error
    // we should however always resolve the request end timing
    if (e && e.status) {
      // this ensures any logging / metrics based on ctx.status will recieve the correct status code
      ctx.status = e.status;
    }

    const endTime = (0, _now.now)() - ctx.timing.start;
    end.resolve(endTime);
    throw e;
  });
}

var _default = (0, _createPlugin.createPlugin)({
  provides: () => timing,
  middleware: () => middleware
});

exports.default = _default;

function deferred() {
  let resolve = () => {};

  let reject = () => {};

  const promise = new Promise((res, rej) => {
    resolve = res;
    reject = rej;
  });
  return {
    promise,
    resolve,
    reject
  };
}
